### PR TITLE
Added waf_alert.py

### PR DESCRIPTION
**Pull request by ishanip24**
## Summary:
This script does a very simplistic check for WAF attack and notifies us via
slack if it notices anything that looks like a WAF attack.

Issue: INFRA-6453

## Test plan:
Run 'python waf_alert.py' and see if there is the appropriate test alert message in the bot-testing Slack channel.

### CODE REVIEW

File: gae_dashboard/waf_alert.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+TABLE_FORMAT = '%Y%m%d'
+TS_FORMAT = '%Y-%m-%d %H:%M:%S'
+
+
+ALERT_CHANNEL = '#bot-testing'
+
+# The fastly's timestamp field is a string with extra details denoting the time
+# zone. BigQuery doesn't understand that part, so we trim that out as the time
+# zone is always +0000.
+today = datetime.datetime.now()
```
Comment (dat-boris) [inline gae_dashboard/waf_alert.py:38]: 2 comments:

1. due to the fact that we might pass in the date from the command line `sys.argv[1]` - we might not always use today.
2. we have to pay extra attention if the date is across date boundary (see dos_alert.py https://github.com/Khan/internal-webserver/blob/master/gae_dashboard/dos_alert.py#L204 for example)
  - Note that there's a comment that we "have" to use legacy SQL to use table decorator.  This is not true anymore as we can use table partition and then `UNION` or table suffix (see https://cloud.google.com/bigquery/docs/querying-wildcard-tables)
  
 This is a technical detail that we can pair on next time we work on this.

File: gae_dashboard/waf_alert.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+waf_log_name = BQ_PROJECT + '.' + FASTLY_DATASET + '.' + FASTLY_WAF_LOG_TABLE_PREFIX + '_' + d1
+log_name = BQ_PROJECT + '.' + FASTLY_DATASET + '.' + FASTLY_LOG_TABLE_PREFIX + '_' + d1
+
+QUERY_TEMPLATE = """\
+#standardSQL
+WITH BLOCKED_REQ AS (
+  SELECT
+    TIMESTAMP_TRUNC(TIMESTAMP(timestamp), MINUTE) AS t,
+    count(*) AS blocked,
+    APPROX_TOP_COUNT(waf.message,1)[OFFSET(0)].value AS message,
```
Comment (dat-boris) [inline gae_dashboard/waf_alert.py:49]: This query does not do what we intend to do.

We would like to get the "top message over the last 5 mins" - while the current query will get the "top messages per minute over the last 5 minutes".

For example: suppose following CSV
```
Time, message, unique_messages
t=1, Message A, 1000
t=2, Message B, 2
t=3, Message B, 2
```
Currently, I believe we will return the aggregate result of:
```
Message B, 1004
```
Which is not what we want.  We should change this to aggregate the information over last 5 mins instead.

**Review by dat-boris — COMMENTED** [review]
Thank you - this looks like a great start!  Thank you for tidying up the code and making it very easy to understand.

I'm commenting on this closed PR since it probably should be reopened (see my comment on https://github.com/Khan/internal-webserver/pull/6#issuecomment-881634775)

The main focus of my comment is mainly on the correctness of the current query.  I have some other comments about the technicality of the code, but those can be addressed later.

File: gae_dashboard/waf_alert.py (hunk 2 above)
Comment (ishanip24) [reply gae_dashboard/waf_alert.py:49]: I was thinking of changing line 47 from
TIMESTAMP_TRUNC(TIMESTAMP(timestamp), MINUTE) AS t 
to 
TIMESTAMP(timestamp) AS t

When I test this, now, I have noticed that the blocked WAF # has drastically dropped. For example:
```
TEST Possible WAF alert
   0.0% (29 / 1341107) of requests blocked in the last 5 minutes
   IP: 98.227.5.102 (27.6% of blocks)
   Request User Agent: Mozilla/5.0 (compatible; SemrushBot/7~bl; +http://www.semrush.com/bot.html) (27.6% of blocks)
   WAF Block Reason: XSS Attack Detected via libinjection (58.6% of blocks)
```

Is this a more accurate depiction of what we expect to see?

**Pull request merged**